### PR TITLE
[Reviewer: Andy] Switch from zc.buildout to a purely setup.py-based build infrastructure (ready to merge)

### PR DESCRIPTION
**Pull request by rkd-msw**
Sending to you for review as our pseudo-buildmaster - should be uncontroversial, as it's mostly a copy/paste from existing setup.py-based Python Makefiles.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -2,8 +2,8 @@
 	path = build-infra
 	url = git@github.com:Metaswitch/clearwater-build-infra.git
 [submodule "modules/common"]
-	path = modules/common
+	path = common
 	url = git@github.com:Metaswitch/python-common.git
 [submodule "modules/telephus"]
-	path = modules/telephus
+	path = telephus
```
Comment (bossmc) [inline .gitmodules]: What are the ramifications of this?  I assume people need to `git submodule sync` at least?

File: Makefile [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
-DEB_NAMES := homer homestead-prov
+DEB_NAMES := crest
 
-MAX_LINE_LENGTH ?= 99
-
-# As we plan to deploy on 64 bit systems, by default target 64 bit. Disable this to attempt to build on 32 bit
-# Note we do not plan to support 32 bit going forward, so this may be removed in the future
-X86_64_ONLY=1
+# The build has been seen to fail on Mac OSX when trying to build on i386. Enable this to build for x86_64 only
+X86_64_ONLY=0
```
Comment (bossmc) [inline Makefile]: Why this change?

File: Makefile [hunk 3]
```diff
@@ -1,15 +1,13 @@
 ENV_DIR := $(shell pwd)/_env
+ENV_PYTHON := ${ENV_DIR}/bin/python
 PYTHON_BIN := $(shell which python)
 
 DEB_COMPONENT := crest
 DEB_MAJOR_VERSION := 1.0${DEB_VERSION_QUALIFIER}
-DEB_NAMES := homer homestead-prov
+DEB_NAMES := crest
```
Comment (bossmc) [inline Makefile]: Huh?

File: Makefile [hunk 4]
```diff
@@ -18,56 +16,58 @@ all: help
...
-	PYTHONPATH=src bash -c 'bin/python src/metaswitch/crest/main.py'
+	@cat docs/development.md
 
 .PHONY: test
-test: bin/python setup.py
-	bin/python setup.py test
+test: setup.py env
+	PYTHONPATH=src:common ${ENV_DIR}/bin/python setup.py test -v
+
+${ENV_DIR}/bin/flake8: env
```
Comment (bossmc) [inline Makefile]: According to @mbt-msw you shouldn't have a real thing depend on a `.PHONY` thing (as `.PHONY` messes with error detection in weird and unexpected ways).  Instead, depend on the sentinel file.

File: Makefile [hunk 5]
```diff
@@ -18,56 +16,58 @@ all: help
...
-	mkdir -p bin/
-	ln -s $(ENV_DIR)/bin/buildout bin/
-
-$(ENV_DIR)/bin/python:
-	virtualenv --no-site-packages --setuptools --python=$(PYTHON_BIN) $(ENV_DIR)
+	
+${ENV_DIR}/.eggs_installed : $(ENV_DIR)/bin/python $(shell find src/metaswitch -type f -not -name "*.pyc") $(shell find common/metaswitch -type f -not -name "*.pyc")
+	# Generate .egg files for crest and python-common
+	${ENV_DIR}/bin/python setup.py bdist_egg -d .eggs
+	cd common && EGG_DIR=../.eggs make build_common_egg
```
Comment (bossmc) [inline Makefile]: Why not `bdist_egg`?  Is `common` not a valid python egg source directory?

File: src/metaswitch/crest/test/api/homestead/provisioning/handlers/private.py [hunk 6]
```diff
@@ -100,7 +100,7 @@ def test_put_plaintext_password(self, put_digest):
     def test_get_missing_subscriber(self, get_digest, send_error):
         get_digest.side_effect = NotFoundException()
         self.handler.get("private_id")
-        get_digest.assert_called
+        self.assertTrue(get_digest.called > 0)
```
Comment (bossmc) [inline src/metaswitch/crest/test/api/homestead/provisioning/handlers/private.py]: These seem like unrelated changes... they look fine but did you mean to include them?

**Comment by bossmc** [comment]
Mostly fine, some comments throughout (some of which may well also apply to our other projects).

File: src/metaswitch/crest/test/api/homestead/provisioning/handlers/private.py (hunk 6 above)
Comment (rkd-msw) [inline src/metaswitch/crest/test/api/homestead/provisioning/handlers/private.py]: I've pulled in an updated Mock as part of this (1.3 vs. 1.0.1) and made the necessary tweaks.

File: Makefile (hunk 5 above)
Comment (rkd-msw) [inline Makefile]: Building python-common requires appropriate compiler flags (to pull in cpp-common for cffi) - I've wrapped that ion a Makefile target.

File: Makefile (hunk 4 above)
Comment (rkd-msw) [inline Makefile]: Yes, this is a problem in our other projects - I'll fix up.

Comment (rkd-msw) [inline Makefile]: Actually, no, I'll raise an issue but not let it hold up this PR.

File: Makefile (hunk 5 above)
Comment (bossmc) [inline Makefile]: Ok, makes sense.

File: Makefile (hunk 2 above)
Comment (rkd-msw) [inline Makefile]: Too much copy-pasting. I propose we just support x86_64 in all Python repos for consistency - agreed?

File: .gitmodules (hunk 1 above)
Comment (rkd-msw) [inline .gitmodules]: Yep - I'll circulate an email when I merge.

File: Makefile (hunk 3 above)
Comment (rkd-msw) [inline Makefile]: Straight copy-paste-replace error.

File: Makefile (hunk 2 above)
Comment (bossmc) [inline Makefile]: I can see no good reason to limit ourselves either way, but if we have to then so be it (`x86_64`).

**Comment by rkd-msw** [comment]
I've tested, including confirming that bulk_create.py works and making a test call with the live tests.
